### PR TITLE
Use JWT and UserProfile classes provided by rain-api-core.auth

### PR DESCRIPTION
**Pull request by Jlrine2**

### CODE REVIEW

**Comment by Jlrine2** [comment]
created an [issue](https://github.com/ASFHyP3/mkdocs-asf-theme/issues/27) in mkdocs-asf-theme for the security vulnerability in mkdocs-material. we can pin the version to 7.1.6 ourselves too if we want

File: lambda/app.py [hunk 1]
```diff
@@ -829,37 +815,30 @@ def dynamic_url():
...
-            log_context(user_id=user_id)
-            log.debug(f'User {user_id} has user profile: {user_profile}')
-            jwt_payload = user_profile_2_jwt_payload(user_id, token, user_profile)
-            log.debug(f"Encoding JWT_PAYLOAD: {jwt_payload}")
-            custom_headers.update(make_set_cookie_headers_jwt(jwt_payload, '', os.getenv('COOKIE_DOMAIN', '')))
-            cookievars[JWT_COOKIE_NAME] = jwt_payload
+                return user_profile
+
+            log_context(user_id=user_profile.user_id)
+            log.debug(f'User {user_profile.user_id} has user profile: {user_profile.to_jwt_payload()}')
```
Comment (reweeden) [inline lambda/app.py:825]: Use the `%s` in log message strings and pass the values as args

File: lambda/app.py [hunk 2]
```diff
@@ -829,37 +815,30 @@ def dynamic_url():
...
+            custom_headers.update(
+                JWT_MANAGER.get_header_to_set_auth_cookie(user_profile, os.getenv('COOKIE_DOMAIN', '')))
```
Comment (reweeden) [inline lambda/app.py:827]:
```suggestion
            custom_headers.update(
                JWT_MANAGER.get_header_to_set_auth_cookie(user_profile, os.getenv('COOKIE_DOMAIN', ''))
            )
```

File: tests/test_app.py [hunk 3]
```diff
@@ -505,53 +524,34 @@ def test_root(mock_get_urs_url, mock_make_html_response, client):
...
+    mock_get_profile.return_value = user_profile
 
     client.http.get("/")
 
     mock_get_urs_url.assert_not_called()
     mock_make_html_response.assert_called_once_with(
         {
             "title": "Welcome",
-            "profile": {"urs-user-id": "user_name"}
+            "profile": user_profile.to_jwt_payload()
```
Comment (reweeden) [inline tests/test_app.py:545]: I'd prefer having a hardcoded dict here just in case the behavior of `to_jwt_payload` changes.

**Review by reweeden — CHANGES_REQUESTED** [review]
Nice! I like that ratio of additions to deletions

**Review by benbart — APPROVED** [review]

File: lambda/app.py [hunk 4]
```diff
@@ -82,16 +78,27 @@ def wrapper(*args, **kwargs):
 
 STAGE = os.getenv('STAGE_NAME', 'DEV')
 
+JWT_COOKIE_NAME = 'asf-urs'
+JWT_KEYS = retrieve_secret(os.getenv('JWT_KEY_SECRET_NAME'))
```
[the remark this answers is not in the document]
Comment (Jlrine2) [reply lambda/app.py:82]: I agree, just wasn't sure how I wanted to go about that.

File: lambda/app.py [hunk 5]
```diff
@@ -82,16 +78,27 @@ def wrapper(*args, **kwargs):
...
+AUTH_CONFIG = {
+    'algorithm': os.getenv('JWT_ALGO', 'RS256'),
+    'public_key': base64.b64decode(JWT_KEYS.get('rsa_pub_key', '')).decode(),
+    'private_key': base64.b64decode(JWT_KEYS.get('rsa_priv_key', '')).decode(),
+    'cookie_name': os.getenv('JWT_COOKIENAME', '')
+}
+JWT_MANAGER = JwtManager(**AUTH_CONFIG)
```
[the remark this answers is not in the document]
Comment (Jlrine2) [reply lambda/app.py:89]: good catch

File: lambda/app.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (Jlrine2) [reply lambda/app.py:83]: We could set expensive globals to `None` and then call a function at the beginning of `TeaChalice.__call__` that updates them using `ttl_cache` to save on performance. Not sure how I feel about that tho

File: lambda/app.py [hunk 6]
```diff
@@ -82,16 +79,37 @@ def wrapper(*args, **kwargs):
...
+    cookie_name=os.getenv('JWT_COOKIENAME', '')
+)
+
+
+@ttl_cache(maxsize=2, ttl=10*60, timer=time.time)
+def get_black_list():
+    endpoint = os.getenv('BLACKLIST_ENDPOINT', '')
+    if endpoint:
+        response = urllib.request.urlopen(endpoint).read().decode('utf-8')
+        return json.loads(response)['blacklist']
```
Comment (reweeden) [inline lambda/app.py:98]: Do we need to do any error handling in case the blacklist file is malformed?

File: tests/test_app.py [hunk 7]
```diff
@@ -505,53 +531,43 @@ def test_root(mock_get_urs_url, mock_make_html_response, client):
...
     client.http.get("/")
 
     mock_get_urs_url.assert_not_called()
     mock_make_html_response.assert_called_once_with(
         {
             "title": "Welcome",
-            "profile": {"urs-user-id": "user_name"}
+            "profile": {
+                'urs-user-id': 'test_user',
+                'urs-access-token': 'test_token',
```
Comment (reweeden) [inline tests/test_app.py:554]: Ahhh, half are single and half are double, pls no.

**Review by benbart — APPROVED** [review]

**Pull request merged**